File: src/main/java/raccoonlang/parser/ast/FqtnAstNode.java
package raccoonlang.parser.ast;

import raccoonlang.parser.Parser;
import raccoonlang.tokenizer.Token;
import raccoonlang.tokenizer.TokenType;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FqtnAstNode {

    List<Token> identifiers = new ArrayList<>();

    public static FqtnAstNode parse(Parser parser) {
        FqtnAstNode fqtnAstNode = new FqtnAstNode();

        fqtnAstNode.identifiers.add(parser.take(TokenType.IDENTIFIER));

        while(parser.peek().getTokenType() == TokenType.DOT){
            parser.take();
            fqtnAstNode.identifiers.add(parser.take(TokenType.IDENTIFIER));
        }

        //TODO: generics

        return fqtnAstNode;
    }

    @Override
    public String toString() {
        return "FqtnAstNode{" +
                "identifiers=" + identifiers.stream().map(x -> x.getText()).reduce((a, b) -> a + "." + b).get() +
                '}';
    }
}

File: Parsing/AST/FqtnAstNode.cs
namespace Raccoonlang.Parsing.AST;

using Tokenizing;

// fully qualified type name
public class FqtnAstNode 
{
    public List<Token> Identifiers = new List<Token>();
    public GenericTypesAstNode? Node;

    public static FqtnAstNode Parse(Parser parser)
    {
        FqtnAstNode node = new FqtnAstNode();

        node.Identifiers.Add(parser.Take(TokenType.IDENTIFIER));

        while(parser.Peek().Type == TokenType.DOT) {
            parser.Take();
            node.Identifiers.Add(parser.Take(TokenType.IDENTIFIER));
        }

        node.Node = GenericTypesAstNode.TryParse(parser);
        if (node.Node == null) Console.WriteLine("Could not find a generic here! This might be right, or wrong, check source!");
        return node;
    }

    public override string ToString() {
        return "FqtnAstNode{identifiers=" + string.Join(",", Identifiers) + "}";
    }
}